### PR TITLE
feat(api/build): add endpoint to get build by id

### PR DESCRIPTION
**Pull request by ecrupper**
Getting build by ID will be very helpful in a number of ways, but the reason at the top of mind is using this endpoint to retrieve a build that corresponds with a hook on the hooks page in the UI.

One endpoint is for users who have at least `read` access to the repo in question, and the other is for admins which bypasses that repo check.

Part of the effort for https://github.com/go-vela/community/issues/610 as well.

### CODE REVIEW

File: router/search.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+import (
+	"github.com/gin-gonic/gin"
+	"github.com/go-vela/server/api"
+)
+
+// SearchHandlers is a function that extends the provided base router group
+// with the API handlers for source code management functionality.
+//
+// GET    /api/v1/search/build/:id .
```
Comment (jbrockopp) [inline router/search.go:15]: plural?

**Review by jbrockopp — CHANGES_REQUESTED** [review]

**Review by jbrockopp — CHANGES_REQUESTED** [review]
I've reviewed the code and it LGTM! However, I had a question that may lead to a discussion.

> NOTE: I'm posting review as request for changes for visibility purposes.

I noticed that today, to query a build for a repo, we require the user has at least `read` level access:

https://github.com/go-vela/server/blob/8e8d0d456b0f57a978e996b85a20402841fabc5e/router/build.go#L56

Did we intentionally make this endpoint a higher access level (`write`)?

https://github.com/go-vela/server/blob/1de361fa72c98423be23ad3f3e7648b0800e1406/api/build.go#L492-L501

If so, could you share some of the thinking around why?

FWIW - I don't have strong concerns about the difference in access level requirements between the two endpoints.

More so, I'm trying to understand the reasoning behind it.

**Comment by ecrupper** [comment]
Yeah, I'm not really sure why I chose `write`. I think complete parity with the current form of getting a build is the way we want to go. I'll change that now

**Review by jbrockopp — APPROVED** [review]
LGTM

File: api/admin/build.go [hunk 2]
```diff
@@ -20,6 +20,55 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+// swagger:operation GET /api/v1/admin/builds/:id admin AdminSingleBuild
```
Comment (wass3r) [inline api/admin/build.go:23]: with the latest changes, this needs to be updated to `/api/v1/admin/build/:id`

**Review by wass3r — CHANGES_REQUESTED** [review]

File: api/build.go [hunk 3]
```diff
@@ -411,6 +411,96 @@ func skipEmptyBuild(p *pipeline.Build) string {
...
+func GetBuildByID(c *gin.Context) {
+	// Variables that will hold the library types of the build and repo
+	var (
+		b *library.Build
+		r *library.Repo
+	)
+
+	// Capture user from middleware
+	u := user.Retrieve(c)
+
```
Comment (plyr4) [inline api/build.go:454]: can we add a simple log here like `GetBuild` ?
https://github.com/go-vela/server/blob/c565a1d762bbd835b4f05f3d7318b9eda296757d/api/build.go#L955-L963

File: api/build.go [hunk 4]
```diff
@@ -411,6 +411,96 @@ func skipEmptyBuild(p *pipeline.Build) string {
...
+	}
+
+	// Capture user access from SCM
+	perm, err := scm.FromContext(c).RepoAccess(u, u.GetToken(), r.GetOrg(), r.GetName())
+	if err != nil {
+		logrus.Errorf("unable to get user %s access level for repo %s", u.GetName(), r.GetFullName())
+	}
+
+	// Ensure that user has at least read access to repo to return the build
+	if perm == "none" {
```
[the remark this answers is not in the document]
Comment (ecrupper) [reply api/build.go:493]: The middleware won't capture the repo from the request, so it's necessary to have this perm check after retrieving the build. This preserves the integrity of the permission barrier.

File: api/admin/build.go [hunk 5]
```diff
@@ -20,6 +20,55 @@ import (
...
+//     schema:
+//       "$ref": "#/definitions/Build"
+//   '500':
+//     description: Unable to retrieve build from the database
+//     schema:
+//       "$ref": "#/definitions/Error"
+
+// SingleBuild represents the API handler to
+// capture a single build stored in the database by id.
+func SingleBuild(c *gin.Context) {
```
[the remark this answers is not in the document]
Comment (ecrupper) [reply api/admin/build.go:44]: There used to be an admin endpoint called `AllBuilds`, so I thought I'd follow that pattern. Now that it's no longer a pattern, I can change this.

**Review by plyr4 — APPROVED** [review]

**Review by wass3r — APPROVED** [review]

**Pull request merged**